### PR TITLE
LPD-16060: Allow to PATCH Object Entries External Reference Code using REST API

### PR DESCRIPTION
**Pull request by sergiojimcos**
## Motivation
The idea of this PR is to allow users to PATCH Object Entries External Reference Code using REST APIs. This operation is supported for PATCH Object Entries by Id and `External Reference Code`

**What is new?**
- Allow to Patch Object Entries External Reference Codes in `ObjectEntryManager` implementations.
- Add test cases for that.

**Steps to reproduce**
1. Create a Custom Object Entry called Student and publish it
2. Create a Student Object Entry with ERC: `Student1`
3. Execute this query:

```curl
curl -X 'PATCH' \
  'http://localhost:8080/o/c/students/by-external-reference-code/Student1' \
  -H 'accept: application/json' \
  -H 'Content-Type: application/json' \
  -u 'test@liferay.com:test' \
  -d '{
"externalReferenceCode": "NewStudent1"
}'
```

**After PR**
The External Reference Code will be updated with the new one: `NewStudent1`

**Before PR**
The External Reference Code is not being updated.

**JIRA Related Tickets**
https://liferay.atlassian.net/browse/LPD-16060

### CODE REVIEW

**Comment by sergiojimcos** [comment]
ci:test:relevant

File: modules/apps/object/object-rest-test/src/testIntegration/java/com/liferay/object/rest/internal/resource/v1_0/test/ObjectEntryResourceTest.java [hunk 1]
```diff
@@ -6780,6 +6817,57 @@ private void _testGetNestedFieldDetailsInRelationships(
...
+				JSONUtil.put(
+					_OBJECT_FIELD_NAME_1, RandomTestUtil.randomString()
+				).put(
+					"externalReferenceCode", _ERC_VALUE_2
+				).toString(),
+				objectDefinitionRESTContextPath, Http.Method.POST));
+
+		Assert.assertEquals(
+			400,
+			HTTPTestUtil.invokeToHttpCode(
```
Comment (ccorreagg) [inline modules/apps/object/object-rest-test/src/testIntegration/java/com/liferay/object/rest/internal/resource/v1_0/test/ObjectEntryResourceTest.java:6800]: Furthermore, we would still have the LPP issue. Please, read the LPP and talk to PM to decide which is the expected behaviour when a user send a PUT request with different ERC. It is not clear to me

**Review by ccorreagg — CHANGES_REQUESTED** [review]
Please, review my comments

File: modules/apps/object/object-rest-api/src/main/java/com/liferay/object/rest/manager/v1_0/ObjectEntryManager.java [hunk 2]
```diff
@@ -60,6 +60,11 @@ public default ObjectEntry partialUpdateObjectEntry(
 			companyId, dtoConverterContext, externalReferenceCode,
 			objectDefinition, scopeKey);
 
+		if (objectEntry.getExternalReferenceCode() != null) {
```
[the remark this answers is not in the document]
Comment (sergiojimcos) [reply modules/apps/object/object-rest-api/src/main/java/com/liferay/object/rest/manager/v1_0/ObjectEntryManager.java:63]: Make sense!!

File: modules/apps/object/object-rest-test/src/testIntegration/java/com/liferay/object/rest/internal/resource/v1_0/test/ObjectEntryResourceTest.java (hunk 1 above)
Comment (sergiojimcos) [reply modules/apps/object/object-rest-test/src/testIntegration/java/com/liferay/object/rest/internal/resource/v1_0/test/ObjectEntryResourceTest.java:6800]: hey @ccorreagg! I've been talking with PM and after making some tests in Object Entries PUT operations, we found that it updates the ERC using the byId and byERC endpoints. For that reason, we should able to PATCH Object Entries ERC byId and byERC endpoints. You can check it in the LPD description, PM has updated it

**Comment by sergiojimcos** [comment]
ci:test:sf

**Comment by sergiojimcos** [comment]
ci:test:relevant

File: modules/apps/object/object-rest-test/src/testIntegration/java/com/liferay/object/rest/internal/resource/v1_0/test/ObjectEntryResourceTest.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (sergiojimcos) [reply modules/apps/object/object-rest-test/src/testIntegration/java/com/liferay/object/rest/internal/resource/v1_0/test/ObjectEntryResourceTest.java:6800]: As I have mentioned in the next comment, PM has updated the requirements for this bug. You can check it in the description

File: modules/apps/object/object-rest-test/src/testIntegration/java/com/liferay/object/rest/internal/resource/v1_0/test/ObjectEntryResourceTest.java [hunk 3]
```diff
@@ -6780,6 +6817,57 @@ private void _testGetNestedFieldDetailsInRelationships(
...
+		throws Exception {
+
+		JSONObject jsonObject = HTTPTestUtil.invokeToJSONObject(
+			JSONUtil.put(
+				"externalReferenceCode", externalReferenceCode
+			).toString(),
+			endpoint, Http.Method.PATCH);
+
+		Assert.assertEquals(
+			jsonObject.getString("externalReferenceCode"),
```
Comment (ccorreagg) [inline modules/apps/object/object-rest-test/src/testIntegration/java/com/liferay/object/rest/internal/resource/v1_0/test/ObjectEntryResourceTest.java:6770]: I think the first parameter should be the expected value to match it with the argument names:

`public static void assertEquals(Object expected, Object actual)`

File: modules/apps/object/object-rest-test/src/testIntegration/java/com/liferay/object/rest/internal/resource/v1_0/test/ObjectEntryResourceTest.java [hunk 4]
```diff
@@ -8240,6 +8178,42 @@ private void _testPutCustomObjectEntryUnlinkNestedCustomObjectEntries(
 		}
 	}
 
+	private void _testPutCustomObjectEntryWithDuplicateExternalReferenceCode(
+			String objectDefinitionRESTContextPath)
```
Comment (ccorreagg) [inline modules/apps/object/object-rest-test/src/testIntegration/java/com/liferay/object/rest/internal/resource/v1_0/test/ObjectEntryResourceTest.java:8182]: This parameter is called different to provide the same information in the `_testPatchCustomObjectEntryWithDuplicateExternalReferenceCode` method. Please unify with the same name!

File: modules/apps/object/object-rest-test/src/testIntegration/java/com/liferay/object/rest/internal/resource/v1_0/test/ObjectEntryResourceTest.java [hunk 5]
```diff
@@ -8240,6 +8178,42 @@ private void _testPutCustomObjectEntryUnlinkNestedCustomObjectEntries(
 		}
 	}
 
+	private void _testPutCustomObjectEntryWithDuplicateExternalReferenceCode(
```
Comment (ccorreagg) [inline modules/apps/object/object-rest-test/src/testIntegration/java/com/liferay/object/rest/internal/resource/v1_0/test/ObjectEntryResourceTest.java:8181]: This method is almost equal to the `_testPatchCustomObjectEntryWithDuplicateExternalReferenceCode`:

Could you create just one? See this pattern: https://github.com/liferay/liferay-portal/blob/e9dee38587df6fe28b484f23cb5ae994e807aa47/modules/apps/object/object-rest-test/src/testIntegration/java/com/liferay/object/rest/internal/resource/v1_0/test/ObjectEntryResourceTest.java#L4738

image

File: modules/apps/object/object-rest-test/src/testIntegration/java/com/liferay/object/rest/internal/resource/v1_0/test/ObjectEntryResourceTest.java [hunk 6]
```diff
@@ -5940,73 +5940,12 @@ public void testPutCustomObjectEntryUnlinkNestedCustomObjectEntriesByExternalRef
 	public void testPutCustomObjectEntryWithDuplicateExternalReferenceCode()
```
Comment (ccorreagg) [inline modules/apps/object/object-rest-test/src/testIntegration/java/com/liferay/object/rest/internal/resource/v1_0/test/ObjectEntryResourceTest.java:5940]: If there is a test for the PUT and another one for the PATCH when the external reference code is duplicated, I miss a test for updating the ERC in the PUT endpoint

testPatchCustomObjectEntryExternalReferenceCode vs testPutCustomObjectEntryExternalReferenceCode

Could we include the test in this PR?

**Review by ccorreagg — CHANGES_REQUESTED** [review]
Please review my comments

File: modules/apps/object/object-rest-test/src/testIntegration/java/com/liferay/object/rest/internal/resource/v1_0/test/ObjectEntryResourceTest.java (hunk 6 above)
Comment (sergiojimcos) [reply modules/apps/object/object-rest-test/src/testIntegration/java/com/liferay/object/rest/internal/resource/v1_0/test/ObjectEntryResourceTest.java:5940]: Sure!!

**Comment by sergiojimcos** [comment]
ci:test:relevant

**Review by ccorreagg — APPROVED** [review]

**Comment by sergiojimcos** [comment]
After taking a view at the test results, I've analysed them and they are not related

**Comment by sergiojimcos** [comment]
ci:foward:force

**Comment by sergiojimcos** [comment]
Sent to Brian here: https://github.com/brianchandotcom/liferay-portal/pull/147070

**Pull request closed without merging**